stronk/business/program.py

```diff
@@ -0,0 +1,75 @@
+from stronk.models.program import Program as ProgramModel
+from stronk.models.program_workouts import ProgramWorkouts as ProgramWorkoutsModel
+from stronk.models.workout_exercise import WorkoutExercise as WorkoutExerciseModel
+from stronk.models.workout import Workout as WorkoutModel
+from stronk.schemas.user.type import User
+
+
+def cloneProgram(program_id: int, user_id: int):
+    """duplicates a program and nested relations with the `user_id` as the author
+
+    args:
+        program_id (int): id of the program to be cloned
+        user_id (int): id of the user designated to be the author of the cloned models
+    """
+    old_program = ProgramModel.find_by_id(program_id)
+    if not old_program:
+        raise NotFound(PROGRAM_NOT_FOUND_MSG)
+
+    new_program = old_program.clone(user_id)
+
+    # clone related relations
+    old_program_workouts = ProgramWorkoutsModel.filter_by_program_id(old_program.id)
+    for old_program_workout in old_program_workouts:
+
+        # create new workout
+        old_workout = WorkoutModel.find_by_id(
+            old_program_workout.workout_id)
+        if not old_workout:
+            raise NotFound(WORKOUT_NOT_FOUND_MSG)
+
+        new_workout = old_workout.clone(user_id)
+        # create new program workout with the new program id and workout
+        new_program_workout = old_program_workout.clone(
+            new_program_id=new_program.id,
+            new_workout_id=new_workout.id
+        )
+
+        old_workout_exercises = WorkoutExerciseModel.find_workout_exercises(
+            workout_id=old_program_workout.workout_id)
+        for old_workout_exercise in old_workout_exercises:
+            old_workout_exercise.clone(new_workout_id=new_workout.id)
+
+    return new_program
+
+
+def deletePrograms(user_id: int):
+    """deletes all programs (and related nested models) created by the user
+    note: this will not delete exercises created by the user
+    """
+    authored_programs = ProgramModel.find_by_author(user_id)
+    for program in authored_programs:
+        # delete related relations that have the user as their author
+        program_workouts = ProgramWorkoutsModel.filter_by_program_id(program.id)
+
+        for program_workout in program_workouts:
+
+            workout = WorkoutModel.find_by_id(program_workout.workout_id)
+
+            # delete workout exercises -> program workouts -> workouts -> programs
+            workout_exercises = WorkoutExerciseModel.find_workout_exercises(workout_id=workout.id)
+            for workout_exercise in workout_exercises:
+                # update exercise author to deleted
+                exercise = ExerciseModel.find_by_id(id=workout_exercise.exercise_id)
+                exercise.author = DELETED_USER_ID
+                # only need to delete workout exercise, not exercise
+                workout_exercise.delete()
+
+            program_workout.delete()
+            workout.delete()
```

Comment:
I understand it is a nested model but is it possible to make this simpler by also having the program id attached to the workout exercise?
We have to go from program -> workout -> workout exercise and I noticed we end up deleting all 3 after updating the base exercise model's author.
Am I missing something or could we also attach the program id to the workout exercise then simply update the exercise model's author by doing a ExerciseModel.find_by_id(id=program_id)? However, I realize if we add a program id field we'd need to update that since the program is being deleted. 

Again, I understand its a nested model and this looks fine but is there an issue with my suggestion that I'm unaware of? 

I believe my suggestion would get rid of one of the nests and just be doing: 

for program in authored_programs: 
    program_workouts = ProgramWorkoutsModel.filter_by_program_id(program.id)
    workout_exercises  =  ExerciseModel.find_by_id(id=program_id);
    for workout_exercise in workout_exercises: 
        updates and delete 
    for program_workout in program_workouts:
       ...
      program_workout.delete()
      workout.delete()

The way I'm seeing this is if the program is deleted then all related workouts to that program is deleted as well but we are keeping the exercises. So since workouts belong to a specific program
and the workout exercises belong to a specific workout id would it be possible to do what I suggested? It would remove a loop but it does add a field to exercise. I am fine with what is being done
right now but do you have any opinions on this?

Reply:
Seems its missing the indentation. 
`for program in authored_programs:
    program_workouts = ProgramWorkoutsModel.filter_by_program_id(program.id) 
    workout_exercises = ExerciseModel.find_by_id(id=program_id);
    for workout_exercise in workout_exercises:
        updates and delete
    for program_workout in program_workouts:
       ....
    program_workout.delete() 
    workout.delete()`

Reply:
Yee don't be afraid of making suggestions :) 
This is a good suggestion! I do agree that flattening it looks a lot better and makes it much more readable but I'm a bit hesitant to the `program_id` field to the `workout _exercise` atm bc it means we'll have to update a bunch of the logic associated with creating `program_exercises` to enable it. I still think we should consider doing it at some point, so do you want to make an issue on the JIRA so that we can tackle this in a later PR?

Reply:
Sounds good to me